### PR TITLE
feat: lemmas characterising List.minimum?

### PR DESCRIPTION
**Pull request by semorrison**
The full characterisation is only given for `List Nat`, as Std doesn't have the necessary order typeclasses to prove it generally. I'm using `List.minimum?` over `Nat`, so this suffices for my purposes for now. There's a comment about generalizing.

### CODE REVIEW

File: Std/Data/List/Lemmas.lean [hunk 1]
```diff
@@ -2167,3 +2167,48 @@ theorem reverse_range' : ∀ s n : Nat, reverse (range' s n) = map (s + n - 1 -
...
+theorem minimum?_cons [Min α] {xs : List α} : (x :: xs).minimum? = foldl min x xs := rfl
+
+@[simp] theorem minimum?_eq_none_iff {xs : List α} [Min α] : xs.minimum? = none ↔ xs = [] := by
+  cases xs <;> simp [minimum?]
+
+-- This could be generalized away from `Nat`.
+-- The standard library doesn't yet have the order typeclasses
+-- that would be necessary for such a generalization.
+theorem minimum?_eq_some_iff {xs : List Nat} :
+    xs.minimum? = some a ↔ (a ∈ xs ∧ ∀ b ∈ xs, a ≤ b) := by
```
Comment (digama0) [inline Std/Data/List/Lemmas.lean:2186]: Could this be proven in general given a plain hypothesis saying the necessary assumptions about `min` and `le` interaction?

Comment (semorrison) [reply Std/Data/List/Lemmas.lean:2186]: Yes, but it's quite a few hypotheses!

`Nat.min_le_left`, `Nat.min_le_right`, `Nat.le_trans`, `Nat.min_def`, `Nat.le_refl`, `Nat.lt_of_le_of_ne`, at least.

Comment (digama0) [reply Std/Data/List/Lemmas.lean:2186]: My guess is that you don't need all of those to prove the theorem, but rather some combination of them. I suspect `le_total` and `le_antisymm` suffice.

Comment (semorrison) [reply Std/Data/List/Lemmas.lean:2189]: I needed:
```
[Min α] [LE α] [DecidableRel ((· : α) ≤ ·)]
    [Trans ((· : α) ≤ ·) (· ≤ ·) (· ≤ ·)] [Antisymm ((· : α) ≤ ·)]
    (le_refl : ∀ a : α, a ≤ a) (le_total : ∀ a b : α, a ≤ b ∨ b ≤ a)
    (le_min_iff : ∀ a b c : α, a ≤ min b c ↔ a ≤ b ∧ a ≤ c)
```

Comment (semorrison) [reply Std/Data/List/Lemmas.lean]: Thanks!

Comment (fgdorais) [reply Std/Data/List/Lemmas.lean]: Note that `le_refl` follows from `le_total`.

Comment (semorrison) [reply Std/Data/List/Lemmas.lean]: (No longer relevant given Mario's refactor.)

**Pull request merged**